platforms/core-execution/execution/README.md

```diff
@@ -0,0 +1,155 @@
+Each unit has two identifiers:
+
+- **Identity**: A locally unique identifier of the work with respect to the current execution scope (e.g., the Gradle build tree).
+- **Cache Key**: A globally unique identifier used to store and retrieve the outputs of the work across time and space.
+
+### Hashing
+
+_Hashing_ is widely used in generating these identifiers.
+We use cryptographic hashing to generate universally unique identifiers for units of work (currently using the MD5 algorithm[^md5-safety][^other-crypto-hashes]).
+While non-cryptographic hash algorithms like [xxHash](https://xxhash.com) or [MurmurHash](https://en.wikipedia.org/wiki/MurmurHash) are significantly faster, they lack the astronomical collision resistance we require.
+
+[^md5-safety]: MD5 has long been compromised from a security standpoint, but our goal is not to protect against malicious intent.
+To avoid accidental collisions, MD5 is still sufficiently strong.
+We use it because it is very fast and universally available on the JVM, and it requires only 16 bytes per hash, which helps conserve memory compared to SHA1 (20 bytes) or SHA256 (32 bytes).
+
+[^other-crypto-hashes]: Other cryptographic hashes could be used, and making the hashing configurable is sensible.
+The [BLAKE family](https://en.wikipedia.org/wiki/BLAKE_(hash_function)) of cryptographic hash functions looks promising for performance, though more research is needed.
+
+We use [Merkle trees](https://en.wikipedia.org/wiki/Merkle_tree) to generate a single hash representing complex inputs.
+This involves hashing together hashes of individual components to create a hash for the whole.
+
+#### Input Normalization
+
+Not all inputs are relevant for a unit of work, and we can exploit this for better performance.
+For example, in a Java compilation task, line endings in source files are irrelevant; whether files use `\n` or `\r\n` does not affect the compilation result.
+Therefore, changes in line endings do not necessitate recompilation, and previously generated `.class` files can be considered up-to-date.
+Ignoring such differences also allows reusing cached compilation results across different operating systems via the remote build cache.
+
+To leverage this, the execution engine allows units of work to define how their inputs should be **normalized**.
+Besides line-ending normalization for source files, Java compilation can also declare the compilation classpath to be normalized via _ABI extraction_.
+This means input hashes are calculated based on the extracted ABI of the classpath rather than raw file contents.
+This allows reuse of existing compilation results despite differences in dependencies, as long as those differences don't affect the ABI.
+
+Input normalization is a key feature that significantly increases the chance of reusing existing results, boosting execution performance.
+
+## Optimizations
+
+### Incremental Build (Up-to-Date Checks)
+
+_TBD_
+
+### The Build Cache
```

Comment:
:x: It's not the part of the execution engine. And it's in another module. While it's ok to be referenced from this doc, its description should be somewhere else (e.g. in the build-cache module or in core-execution/README.md).